lenskit-slim/src/main/java/org/lenskit/slim/CovarianceUpdate.java

```diff
@@ -0,0 +1,196 @@
+     * which means final learned weight should at least make the value of loss function less than the starting value.
+     * @param labels label vector
+     * @param trainingDataMatrix Map of item IDs to item rating vectors.
+     * @return weight vectors learned
+     */
+    @Override
+    public Long2DoubleMap fit(Long2DoubleMap labels, Long2ObjectMap<Long2DoubleMap> trainingDataMatrix) {
+        Long2DoubleMap weights = new Long2DoubleOpenHashMap();
+        final double lambda = updateParameters.getLambda();
+        final double beta = updateParameters.getBeta();
+
+
+        Long2DoubleMap dotProdsOfXsY = new Long2DoubleOpenHashMap();
+        Long2ObjectMap<Long2DoubleMap> innerProdsOfXs = new Long2ObjectOpenHashMap<>();
+        Long2DoubleMap residuals = computeResiduals(labels, trainingDataMatrix, weights);
+
+        // record difference of loss function in two adjacent updating circles
+        double[] loss = {0.0, 0.0};
+        double lossDiff = Double.POSITIVE_INFINITY;
+        TrainingLoopController controller = updateParameters.getTrainingLoopController();
+        int k = 0;
+        while (controller.keepTraining(lossDiff)) {
+            LongIterator items = trainingDataMatrix.keySet().iterator();
+            while (items.hasNext()) {
+                long j = items.nextLong();
+                Long2DoubleMap itemXj = trainingDataMatrix.get(j);
+                double dotProdOfXjY;
+
+                // Storing dot product of column j and labels
+                if (dotProdsOfXsY.containsKey(j)) {
+                    dotProdOfXjY = dotProdsOfXsY.get(j);
+                } else {
+                    dotProdOfXjY = Vectors.dotProduct(itemXj, labels);
+                    dotProdsOfXsY.put(j, dotProdOfXjY);
+                }
+
+                Long2DoubleMap nonzeroWeights = filterValues(weights, 0.0);
+
+                // Storing dot product of column j and column k(s) into correlationOfColumns row j
+                if (innerProdsOfXs.containsKey(j)) {
+                    Long2DoubleMap dotProdsOfXjXks = innerProdsOfXs.get(j);
+                    for (long weightsIdk : nonzeroWeights.keySet()) {
+                        if (!dotProdsOfXjXks.containsKey(weightsIdk)) {
+                            double dotProdOfXjXk = Vectors.dotProduct(itemXj, trainingDataMatrix.get(weightsIdk));
+                            dotProdsOfXjXks.put(weightsIdk, dotProdOfXjXk);
+                        }
+                    }
+                } else {
+                    Long2DoubleMap dotProdsOfXjXks = new Long2DoubleOpenHashMap();
+                    for (long weightsIdk : nonzeroWeights.keySet()) {
+                        double dotProdOfXjXk = Vectors.dotProduct(itemXj, trainingDataMatrix.get(weightsIdk));
+                        dotProdsOfXjXks.put(weightsIdk, dotProdOfXjXk);
+                    }
+                    innerProdsOfXs.put(j, dotProdsOfXjXks);
+                }
```

Comment:
Pulling in an item-item model is just to use its hooks for different similarity strategies. Now the codes for similarity strategies are copied into SLIMBuildContextProvider which will return a mapping from item ids to sets of corresponding neighbors. All the dependencies on KNN are removed.